### PR TITLE
Augmentation refactoring and torchaudio SoX effects support

### PR DESCRIPTION
**Pull request by pzelasko**
TL;DR 

- Changing the data augmentation APIs in Lhotse to accept a callable with signature like: `def augment_fn(audio: Union[torch.tensor, np.ndarray], sampling_rate: int) -> np.ndarray`
- mirroring `WavAugment` capabilities with `torchaudio.sox_effects`

### CODE REVIEW

**Comment by pzelasko** [comment]
It's good to review.

**Comment by pzelasko** [comment]
@freewym please check this out - it will be small but breaking for the recipe you're creating, but hopefully makes the whole setup way easier (no need to compile libsox, wavaugment, etc. just install the latest pytorch + torchaudio with anaconda) and gets us rid of various quirks with multiprocessing.

**Review by mthrok — COMMENTED** [review]
Looks good regarding the usage of torchaudio's Sox Effects.

**Comment by vincentqb** [comment]
(glad to see the migration here! cc facebookresearch/WavAugment#16)

**Comment by freewym** [comment]
> @freewym please check this out - it will be small but breaking for the recipe you're creating, but hopefully makes the whole setup way easier (no need to compile libsox, wavaugment, etc. just install the latest pytorch + torchaudio with anaconda) and gets us rid of various quirks with multiprocessing.

Cool. I think I am supposed to install PyTorch 1.7 in order to test it. I will do later today.

File: lhotse/augmentation/__init__.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import torchaudio as _torchaudio
+
+from .common import AugmentFn
+from .wavaugment import *
+
+if str(_torchaudio.__version__) >= '0.7.0':
```
Comment (mthrok) [inline lhotse/augmentation/__init__.py:6]: FYI: If `torchaudio` hits `0.10.0` release in the future (we do not know if we will move to the major 1.0 release or not), this could produce a wrong result.

```
$ python
>>> '0.7.0' > '0.10.0'
True
```

A future-proof way would be to split the version string and compare major version and minor version as number.

File: lhotse/features/base.py [hunk 2]
```diff
@@ -10,7 +10,7 @@
 import torch
 
 from lhotse.audio import Recording
-from lhotse.augmentation import WavAugmenter
+from lhotse.augmentation import AugmentFn, WavAugmenter
```
Comment (freewym) [inline lhotse/features/base.py:13]: WavAugmenter is no longer useful (?) If it is the case, all appearances of WavAugmenter should be removed within this file

Comment (pzelasko) [reply lhotse/features/base.py:13]: Good point. I want to keep it for now so that people with PyTorch older than 1.7 can also use it, but I will probably add a deprecation warning...

File: lhotse/augmentation/__init__.py (hunk 1 above)
Comment (pzelasko) [reply lhotse/augmentation/__init__.py:6]: Well spotted!

File: lhotse/augmentation/torchaudio.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+def speed(sampling_rate: int) -> List[List[str]]:
+    return [
+        # Random speed perturbation factor between 0.9x and 1.1x the original speed
+        ['speed', RandomValue(0.9, 1.1)],
+        ['rate', sampling_rate],  # Resample back to the original sampling rate (speed changes it)
+    ]
+
+
+def reverb(sampling_rate: int) -> List[List[str]]:
```
Comment (freewym) [inline lhotse/augmentation/torchaudio.py:99]: Is it possible to make such functions more general, in that they can accept more arguments, e.g., the lower/up bound of room sizes can been passed into this function?

File: lhotse/augmentation/torchaudio.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+@dataclass
+class RandomValue:
+    """
+    Represents a uniform distribution in the range [start, end].
+    """
+    start: Union[int, float]
+    end: Union[int, float]
+
+    def sample(self):
+        return random.uniform(self.start, self.end)
```
Comment (freewym) [inline lhotse/augmentation/torchaudio.py:24]: can we use numpy random functions? It may be easier for me to seed it from outside

File: lhotse/augmentation/torchaudio.py (hunk 3 above)
Comment (pzelasko) [reply lhotse/augmentation/torchaudio.py:99]: Sure, I'll change that

File: lhotse/augmentation/torchaudio.py (hunk 4 above)
Comment (pzelasko) [reply lhotse/augmentation/torchaudio.py:24]: Would using this function help (note that it also makes the random cut ID creation deterministic)? 

https://github.com/lhotse-speech/lhotse/blob/master/lhotse/utils.py#L33

Comment (freewym) [reply lhotse/augmentation/torchaudio.py:24]: OK, never mind. I will seed it in my own code.

File: lhotse/augmentation/torchaudio.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            ]
+            for effect in self.effects
+        ]
+
+
+def speed(sampling_rate: int) -> List[List[str]]:
+    return [
+        # Random speed perturbation factor between 0.9x and 1.1x the original speed
+        ['speed', RandomValue(0.9, 1.1)],
+        ['rate', sampling_rate],  # Resample back to the original sampling rate (speed changes it)
```
Comment (freewym) [inline lhotse/augmentation/torchaudio.py:95]: Looks like this line makes the running hang. It works without this line.

File: lhotse/augmentation/__init__.py (hunk 1 above)
Comment (csukuangfj) [reply lhotse/augmentation/__init__.py:6]: `packaging.version` may be helpful in this case.
See https://stackoverflow.com/questions/11887762/how-do-i-compare-version-numbers-in-python

File: lhotse/augmentation/torchaudio.py (hunk 4 above)
Comment (pzelasko) [reply lhotse/augmentation/torchaudio.py:24]: I'll change it to numpy anyway, I guess more people are used to seeding `numpy` than `random` in training loops

File: lhotse/augmentation/torchaudio.py (hunk 5 above)
Comment (pzelasko) [reply lhotse/augmentation/torchaudio.py:95]: 😫

I'll have a look

File: lhotse/augmentation/torchaudio.py (hunk 3 above)
Comment (pzelasko) [reply lhotse/augmentation/torchaudio.py:99]: Actually I'd rather make a follow-up PR later on, as I'm not sure which parameters it makes sense to tweak and how general they should be. If we want to tweak everything it's simpler to just write your own chain...

File: lhotse/augmentation/torchaudio.py (hunk 5 above)
Comment (pzelasko) [reply lhotse/augmentation/torchaudio.py:94]: I was able to replicate the issue and add a unit test that causes it. I submitted the issue to torchaudio here: https://github.com/pytorch/audio/issues/1021

Comment (freewym) [reply lhotse/augmentation/torchaudio.py:94]: BTW, if replacing RandomValue() above with a function  `_get_value(factor)` where `factor` is simply returned, the running hangs as well.

Comment (pzelasko) [reply lhotse/augmentation/torchaudio.py:94]: Is the function defined as a closure (i.e. within another function) and captures some variable outside of its scope? That could explain it... Otherwise, I don't know.

Comment (freewym) [reply lhotse/augmentation/torchaudio.py:94]: No, it's a independent function outside of any other functions.

https://github.com/freewym/espresso/pull/45/commits/e995f0d61ea9777a366a5a2d3b0bbe9eef900169#diff-3724bd15a663fd58b5fc92daa096bc9eb0c0a96a4c070a1a9e999be0682c985aR298

Comment (pzelasko) [reply lhotse/augmentation/torchaudio.py:97]: @freewym some good news, if you create executor like: `ProcessPoolExecutor(..., mp_context=multiprocessing.get_context("spawn"))` it solves the segfault/hanging problem. Could you try? If it works I'll go on and merge this

Comment (pzelasko) [reply lhotse/augmentation/torchaudio.py:97]: Credits to @mthrok for suggesting this

Comment (pzelasko) [reply lhotse/augmentation/torchaudio.py:97]: (to make it clear: it works for me on the grid, on my mac, and in GitHub CI, so it should be okay)

Comment (freewym) [reply lhotse/augmentation/torchaudio.py:97]: Yeah, it works!

Comment (pzelasko) [reply lhotse/augmentation/torchaudio.py:97]: 🎉
